src/compose/ComposeBox.js

```diff
@@ -336,23 +336,6 @@ class ComposeBox extends PureComponent<Props, State> {
     dispatch(sendTypingStop(narrow));
   };
 
-  UNSAFE_componentWillReceiveProps(nextProps: Props) {
```

Comment:
```
ComposeBox: Stop using UNSAFE_componentWillReceiveProps.

This replaces UNSAFE_componentWillReceiveProps with a key prop, so that
we rerender the ComposeBox when we look at a different narrow or
start/stop editing a message.
[...]
```

This isn't correct: it doesn't cause the ComposeBox to rerender in the normal sense (that `render` is called, and any `componentDidUpdate` is called, etc.). The crucial thing that's happening, and that makes this strategy work, is that the old one gets completely torn down and a new one is created. For example, we have this code that sets the initial state when a `ComposeBox` instance is constructed:

```
  state = {
    isMessageFocused: false,
    isTopicFocused: false,
    isFocused: false,
    isMenuExpanded: false,
    topic:
      this.props.initialTopic !== undefined
        ? this.props.initialTopic
        : isTopicNarrow(this.props.narrow)
        ? topicOfNarrow(this.props.narrow)
        : '',
    message: this.props.initialMessage || '',
    selection: { start: 0, end: 0 },
  };
```

That's how `state.topic` and `state.message` end up being set to the desired values when the `key` changes. It's this ["instance field"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields#public_instance_fields) bit, which I think of as a shortcut to setting things in the [constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/constructor). It only runs once during the lifecycle of an instance.

Reply:
Also (and this is more of a nit), the `key` isn't a prop; I think I usually call it the "`key` attribute`. IIRC you can't access it on `this.props`: https://reactjs.org/docs/lists-and-keys.html

Reply:
Here's a nice explanation of the strategy we've chosen, which I linked to in the CZO discussion: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key

Reply:
Thanks for the explanation! Reworded the commit.